# Теория вероятностей — лабораторные работы № 1–4

Единый блокнот. Лабораторные № 1 и № 2 сохранены; работы № 3 и № 4 содержат математические решения и проверку Python.

# Лабораторная работа № 1. Комбинаторика и непосредственное вычисление вероятностей

In [1]:
import scipy.special as sc

## § 1.5. Элементы комбинаторики
---

### Задача 1.8
> В конкурсе по 5 номинациям участвуют 10 кинофильмов. Сколько существует вариантов распределения призов, если по каждой номинации установлены: а) различные призы; б) одинаковые призы?

* Рассмотрим пункт **а)** – призы различаются по номинациям.

Порядок важен, один и тот же фильм может получить несколько призов. Используем размещения с повторениями:

$N_a = \overline{A}_{10}^{5} = 10^5$

In [2]:
10 ** 5

100000

* Рассмотрим пункт **б)** – призы одинаковые.

Порядок не важен, повторения разрешены. Используем сочетания с повторениями:

$N_b = \overline{C}_{10}^{5} = C_{10+5-1}^{5} = C_{14}^{5}$

In [3]:
int(sc.comb(14, 5))

2002

### Задача 1.9
> Сколько существует семизначных чисел, состоящих из цифр 4, 5 и 6, в которых цифра 4 повторяется 3 раза, а цифры 5 и 6 — по 2 раза?

Используем перестановки с повторениями:

$N = P_7(3,2,2) = \frac{7!}{3! \cdot 2! \cdot 2!}$

In [4]:
int(sc.factorial(7) / (sc.factorial(3) * sc.factorial(2) * sc.factorial(2)))

210

## § 1.6. Непосредственное вычисление вероятностей
---

### Задача 1.10
> Буквы Т, Е, И, Я, Р, О написаны на отдельных карточках. Ребенок берет карточки в случайном порядке и прикладывает одну к другой: а) 3 карточки; б) все 6 карточек. Какова вероятность того, что получится слово: а) «ТОР»; б) «ТЕОРИЯ»?

* Рассмотрим событие $A$ – получилось слово **ТОР**.

Общее число способов выбрать и расположить 3 карточки из 6:

$n = A_6^3 = 6 \cdot 5 \cdot 4$

Благоприятный исход один, поэтому

$P(A) = \frac{1}{A_6^3} = \frac{1}{120}$

In [5]:
round(1 / (6 * 5 * 4), 4)

0.0083

* Рассмотрим событие $B$ – получилось слово **ТЕОРИЯ**.

Все 6 карточек можно переставить $6!$ способами. Благоприятный исход один:

$P(B) = \frac{1}{6!} = \frac{1}{720}$

In [6]:
round(1 / sc.factorial(6), 4)

np.float64(0.0014)

### Задача 1.11
> Используя условие примера 1.10, найти вероятность того, что получится слово «АНАНАС», если на отдельных карточках написаны три буквы А, две буквы Н и одна буква С.

* Рассмотрим событие $A$ – получилось слово **АНАНАС**.

Общее число перестановок шести карточек равно $6!$.

Перестановка трех карточек с буквой А и двух карточек с буквой Н не меняет слово, поэтому число благоприятных случаев:

$m = 3! \cdot 2!$

Тогда

$P(A) = \frac{3! \cdot 2!}{6!} = \frac{1}{60}$

In [7]:
round(sc.factorial(3) * sc.factorial(2) / sc.factorial(6), 4)

np.float64(0.0167)

### Задача 1.16
> За круглым столом рассаживаются 5 мужчин и 5 женщин. Найти вероятность того, что: а) никакие два лица одного пола не сядут рядом; б) супруги сядут рядом, если эти мужчины и женщины образуют 5 супружеских пар.

* Рассмотрим событие $A$ – мужчины и женщины сидят поочередно.

Общее число рассадок:

$n = 10!$

Женщины могут занять четные места $5!$ способами, мужчины — нечетные места $5!$ способами, либо наоборот. Поэтому:

$m = 2(5!)^2$

$P(A) = \frac{2(5!)^2}{10!}$

In [8]:
round(2 * sc.factorial(5) ** 2 / sc.factorial(10), 5)

np.float64(0.00794)

* Рассмотрим событие $B$ – каждая супружеская пара сидит рядом.

Пять пар можно переставить $5!$ способами. Внутри каждой пары муж и жена могут поменяться местами двумя способами:

$m = 5! \cdot 2^5$

$P(B) = \frac{5! \cdot 2^5}{10!}$

In [9]:
round(sc.factorial(5) * 2 ** 5 / sc.factorial(10), 5)

np.float64(0.00106)

### Задача 1.17
> В купейный вагон (9 купе по 4 места) семи пассажирам продано 7 билетов. Найти вероятности того, что пассажиры попали: а) в два купе; б) в семь купе; в) в три купе.

Общее число способов выбрать 7 мест из 36:

$n = C_{36}^{7}$

* Рассмотрим событие $A$ – пассажиры попали в **два купе**.

Два купе из девяти выбираются $C_9^2$ способами. Из 8 мест выбранных купе нужно выбрать 7:

$m = C_9^2 \cdot C_8^7$

$P(A) = \frac{C_9^2 \cdot C_8^7}{C_{36}^7}$

In [10]:
round(sc.comb(9, 2) * sc.comb(8, 7) / sc.comb(36, 7), 7)

np.float64(3.45e-05)

* Рассмотрим событие $B$ – пассажиры попали в **семь купе**.

Семь купе из девяти выбираются $C_9^7$ способами. В каждом купе выбирается одно из четырех мест:

$m = C_9^7 \cdot 4^7$

$P(B) = \frac{C_9^7 \cdot 4^7}{C_{36}^7}$

In [11]:
round(sc.comb(9, 7) * 4 ** 7 / sc.comb(36, 7), 5)

np.float64(0.07066)

* Рассмотрим событие $D$ – пассажиры попали в **три купе**.

Три купе из девяти выбираются $C_9^3$ способами.

Возможные распределения семи пассажиров по трем купе:

$4+2+1,\quad 3+3+1,\quad 3+2+2$

Поэтому число благоприятных случаев:

$m = C_9^3 \left(6C_4^4C_4^2C_4^1 + 3C_4^3C_4^3C_4^1 + 3C_4^3C_4^2C_4^2\right)$

$P(D) = \frac{m}{C_{36}^7}$

In [12]:
m = sc.comb(9, 3) * (
    6 * sc.comb(4, 4) * sc.comb(4, 2) * sc.comb(4, 1)
    + 3 * sc.comb(4, 3) * sc.comb(4, 3) * sc.comb(4, 1)
    + 3 * sc.comb(4, 3) * sc.comb(4, 2) * sc.comb(4, 2)
)
round(m / sc.comb(36, 7), 5)

np.float64(0.00773)

# Лабораторная работа № 2. Классическая и геометрическая вероятность

In [1]:
from itertools import permutations, product
from math import factorial

### Задача 1.38
> Пятитомное собрание сочинений расположено на полке в случайном порядке. Какова вероятность того, что книги стоят слева направо в порядке нумерации томов (от 1 до 5)?

Рассмотрим событие $A$ — книги стоят в порядке $1,2,3,4,5$.

Всего существует

$P_5 = 5!$

перестановок пяти томов. Благоприятный вариант только один, поэтому

$P(A)=\frac{1}{5!}=\frac{1}{120}\approx0.00833$.

In [2]:
orders = list(permutations(range(1, 6)))
favorable = [order for order in orders if order == (1, 2, 3, 4, 5)]

p = len(favorable) / len(orders)

print(f'Всего перестановок: {len(orders)}')
print(f'Благоприятных: {len(favorable)}')
print(f'P(A) = {p:.5f}')

Всего перестановок: 120
Благоприятных: 1
P(A) = 0.00833


### Задача 1.43
> Наудачу взятый телефонный номер состоит из 5 цифр. Какова вероятность того, что в нем все цифры: а) различные; б) одинаковые; в) нечетные? Известно, что номер телефона не начинается с цифры ноль.

Всего допустимых пятизначных телефонных номеров:

$n=9\cdot10^4=90000$.

**а)** Все цифры различные. Для первой цифры — 9 вариантов, далее соответственно 9, 8, 7 и 6 вариантов:

$m_a=9\cdot9\cdot8\cdot7\cdot6=27216$,

$P(A)=\frac{27216}{90000}=0.3024$.

**б)** Все цифры одинаковые. Подходят номера $11111,22222,\ldots,99999$:

$m_b=9$,

$P(B)=\frac{9}{90000}=0.0001$.

**в)** Все цифры нечетные. Для каждой позиции возможны цифры $1,3,5,7,9$:

$m_c=5^5=3125$,

$P(C)=\frac{3125}{90000}=\frac{5}{144}\approx0.0347$.

In [3]:
digits = '0123456789'
first_digits = '123456789'

numbers = [
    (first,) + rest
    for first in first_digits
    for rest in product(digits, repeat=4)
]

total = len(numbers)

different = sum(len(set(number)) == 5 for number in numbers)
same = sum(len(set(number)) == 1 for number in numbers)
odd = sum(all(int(digit) % 2 == 1 for digit in number) for number in numbers)

print(f'Всего допустимых номеров: {total}')
print(f'а) Все цифры различные: {different} случаев, P = {different / total:.4f}')
print(f'б) Все цифры одинаковые: {same} случаев, P = {same / total:.4f}')
print(f'в) Все цифры нечетные: {odd} случаев, P = {odd / total:.4f}')

Всего допустимых номеров: 90000
а) Все цифры различные: 27216 случаев, P = 0.3024
б) Все цифры одинаковые: 9 случаев, P = 0.0001
в) Все цифры нечетные: 3125 случаев, P = 0.0347


> **Примечание.** В ответах к учебнику Кремера для пункта 1.43(а) напечатано значение $\frac{1}{27216}$. Оно не соответствует условию задачи: прямой перебор и классическое определение дают $\frac{27216}{90000}=0.3024$.

### Задача 1.48
> В старинной игре в кости необходимо было для выигрыша получить при бросании трех игральных костей сумму очков, превосходящую 10. Найти вероятности: а) выпадения 11 очков; б) выигрыша.

При бросании трех игральных костей имеется

$n=6^3=216$

равновозможных упорядоченных исходов.

**а)** Сумма очков равна 11. Программно переберем все исходы и посчитаем подходящие.

**б)** Для выигрыша сумма должна быть больше 10. Аналогично подсчитаем число исходов с суммой от 11 до 18.

In [4]:
outcomes = list(product(range(1, 7), repeat=3))
total = len(outcomes)

sum_11 = [outcome for outcome in outcomes if sum(outcome) == 11]
win = [outcome for outcome in outcomes if sum(outcome) > 10]

print(f'Всего исходов: {total}')
print(f'а) Сумма 11: {len(sum_11)} случаев, P = {len(sum_11) / total:.3f}')
print(f'б) Сумма больше 10: {len(win)} случаев, P = {len(win) / total:.3f}')

Всего исходов: 216
а) Сумма 11: 27 случаев, P = 0.125
б) Сумма больше 10: 108 случаев, P = 0.500


---
# Лабораторная работа № 3. Вероятности сложных событий. Принцип включений–исключений. Беспорядки

**Цель:** применять формулы дополнения, суммы и произведения вероятностей, подсчитывать объединения множеств и перестановки без неподвижных точек.

**Основные формулы:**

- $P(A\cup B)=P(A)+P(B)-P(A\cap B)$.
- $P(A\cup B\cup C)=P(A)+P(B)+P(C)-P(A\cap B)-P(A\cap C)-P(B\cap C)+P(A\cap B\cap C)$.
- $P(\bar A)=1-P(A)$.
- При независимости: $P(A\cap B)=P(A)P(B)$.
- Число беспорядков: $D_n=n!\sum_{k=0}^{n}(-1)^k/k!$.

У каждой задачи ниже отдельно приведено математическое решение и проверка числа с помощью Python.

In [ ]:
from math import comb, factorial
from fractions import Fraction
from itertools import permutations

def D(n):
    """Число беспорядков (перестановок без неподвижных точек)."""
    return sum((-1)**k * comb(n, k) * factorial(n-k) for k in range(n+1))

assert [D(k) for k in range(7)] == [1, 0, 1, 2, 9, 44, 265]

## § 1.10. Задачи из учебника Кремера (письменная домашняя работа)

### Задача 1.55. Надёжность прибора

**Условие.** Прибор состоит из трёх независимых узлов с надёжностями $0{,}9$, $0{,}95$ и $0{,}8$. Неисправность любого узла выводит весь прибор из строя. Найти вероятность отказа.

**Решение.** Прибор исправен только тогда, когда работают все три узла. По независимости:

$$P(\text{исправен})=0{,}9\cdot0{,}95\cdot0{,}8=0{,}684.$$

Применим противоположное событие:

$$P(\text{отказ})=1-0{,}684=0{,}316.$$

**Ответ:** $0{,}316$.

In [ ]:
p_155 = 1 - Fraction(9,10)*Fraction(95,100)*Fraction(8,10)
print('1.55:', p_155, '=', float(p_155))
assert p_155 == Fraction(79,250)

### Задача 1.58. Три контрольные работы

**Условие.** Вероятности своевременного выполнения трёх работ: $p_1=0{,}6$, $p_2=0{,}5$, $p_3=0{,}8$. Найти вероятность выполнения **ровно двух** и **не менее двух** работ. При вычислении считаем события независимыми.

**Решение.** Обозначим $q_1=0{,}4$, $q_2=0{,}5$, $q_3=0{,}2$. Ровно две работы выполнены в одном из трёх несовместных случаев:

$$P_2=p_1p_2q_3+p_1q_2p_3+q_1p_2p_3$$
$P_2=0{,}6\cdot0{,}5\cdot0{,}2+0{,}6\cdot0{,}5\cdot0{,}8+0{,}4\cdot0{,}5\cdot0{,}8=0{,}46.$$

Все три работы выполнены с вероятностью $P_3=0{,}6\cdot0{,}5\cdot0{,}8=0{,}24$.

Поэтому $P(\text{не менее двух})=P_2+P_3=0{,}46+0{,}24=0{,}70$.

**Ответ:** а) $0{,}46$; б) $0{,}70$.

In [ ]:
p1,p2,p3 = Fraction(6,10),Fraction(5,10),Fraction(8,10)
q1,q2,q3 = 1-p1,1-p2,1-p3
exactly_two = p1*p2*q3 + p1*q2*p3 + q1*p2*p3
at_least_two = exactly_two+p1*p2*p3
print('1.58: ровно 2 =', exactly_two, '; не менее 2 =', at_least_two)
assert (exactly_two,at_least_two)==(Fraction(23,50),Fraction(7,10))

### Задача 1.61. Часы на чистку

**Условие.** Среди $20$ часов $8$ требуют общей чистки. Одновременно выбирают $3$ часа без возвращения. Найти вероятность того, что хотя бы $2$ выбранных часа требуют чистки.

**Решение.** Всего равновозможных наборов $C_{20}^{3}=1140$. Благоприятны случаи: ровно два часа из восьми, требующих чистки, и один из двенадцати остальных; либо все три требуют чистки:

$$P=\frac{C_8^2C_{12}^1+C_8^3}{C_{20}^3}=\frac{28\cdot12+56}{1140}=\frac{98}{285}\approx0{,}34386.$$

**Ответ:** $98/285$.

In [ ]:
p_161=Fraction(comb(8,2)*comb(12,1)+comb(8,3),comb(20,3))
print('1.61:',p_161,'≈',float(p_161));assert p_161==Fraction(98,285)

### Задача 1.64. Карандаши

**Условие.** В коробке $10$ красных, $3$ синих и $7$ жёлтых карандашей. Берут $3$. Найти вероятности: а) все разных цветов; б) все одного цвета.

**Решение.** Всего наборов $C_{20}^3=1140$.

а) Выбираем по одному карандашу каждого цвета:

$$P_a=\frac{C_{10}^1C_3^1C_7^1}{C_{20}^3}=\frac{10\cdot3\cdot7}{1140}=\frac7{38}.$$

б) Суммируем несовместные случаи «все красные», «все синие», «все жёлтые»:

$$P_b=\frac{C_{10}^3+C_3^3+C_7^3}{C_{20}^3}=\frac{120+1+35}{1140}=\frac{13}{95}.$$

**Ответ:** а) $7/38$; б) $13/95$.

In [ ]:
p_164a=Fraction(10*3*7,comb(20,3))
p_164b=Fraction(comb(10,3)+comb(3,3)+comb(7,3),comb(20,3))
print('1.64: разные =',p_164a,'; одинаковые =',p_164b)
assert (p_164a,p_164b)==(Fraction(7,38),Fraction(13,95))

### Задача 1.67. Минимальное число испытаний

**Условие.** Вероятность наступления события в одном испытании $p=0{,}4$. Сколько независимых испытаний требуется, чтобы вероятность хотя бы одного успеха была не меньше $0{,}8704$?

**Решение.** Вероятность не получить событие в одном испытании $q=0{,}6$, в $n$ независимых испытаниях — $q^n$. Требуем:

$$1-0{,}6^n\geq0{,}8704 \quad\Longleftrightarrow\quad 0{,}6^n\leq0{,}1296.$$

Поскольку $0{,}6^4=0{,}1296$, минимальное $n=4$. При $n=3$ вероятность меньше требуемой.

**Ответ:** $4$ испытания.

In [ ]:
p=Fraction(4,10); target=Fraction(8704,10000)
n_167=next(n for n in range(1,100) if 1-(1-p)**n>=target)
print('1.67: n =',n_167);assert n_167==4

### Задача 1.68. Книги

**Условие.** Из $10$ книг $3$ посвящены теории вероятностей. Случайно выбирают $3$ книги. Найти вероятность того, что хотя бы одна из них — по теории вероятностей.

**Решение.** Противоположное событие: все три книги взяты из остальных семи.

$$P(A)=1-\frac{C_7^3}{C_{10}^3}=1-\frac{35}{120}=\frac{17}{24}.$$

**Ответ:** $17/24$.

In [ ]:
p_168=1-Fraction(comb(7,3),comb(10,3))
print('1.68:',p_168);assert p_168==Fraction(17,24)

### Задача 1.71. Вызовы радиста

**Условие.** Радист вызывает корреспондента трижды. Вероятности услышать вызовы: $0{,}2$, $0{,}3$, $0{,}4$; события независимы. Найти вероятность, что корреспондент услышит хотя бы один вызов.

**Решение.** Ни один вызов не услышан с вероятностью

$$P(\text{ни одного})=(1-0{,}2)(1-0{,}3)(1-0{,}4)=0{,}8\cdot0{,}7\cdot0{,}6=0{,}336.$$

Искомое $P=1-0{,}336=0{,}664$.

**Ответ:** $0{,}664$.

In [ ]:
p_171=1-Fraction(8,10)*Fraction(7,10)*Fraction(6,10)
print('1.71:',p_171,'=',float(p_171));assert p_171==Fraction(83,125)

## Принцип включений–исключений — задачи из практики № 3

### Задача 3.1. Сданные экзамены

**Условие.** Из $100$ студентов историю сдали $56$, математику $44$, физику $42$; историю или математику — $76$, историю или физику — $76$, математику или физику — $66$; все три — $5$. Найти: а) сдавших хотя бы один предмет; б) не сдавших ни одного; в) сдавших только физику.

**Решение.** Из двухмножественной формулы получаем $|И\cap М|=56+44-76=24$, $|И\cap Ф|=56+42-76=22$, $|М\cap Ф|=44+42-66=20$.

$$|И\cup М\cup Ф|=56+44+42-24-22-20+5=81.$$

Не сдали ничего $100-81=19$. Только физику $42-22-20+5=5$.

**Ответ:** а) $81$; б) $19$; в) $5$.

In [ ]:
hm,hf,mf=56+44-76,56+42-76,44+42-66
at_least_one=56+44+42-hm-hf-mf+5
none=100-at_least_one
only_physics=42-hf-mf+5
print('3.1:',at_least_one,none,only_physics)
assert (at_least_one,none,only_physics)==(81,19,5)

### Задача 3.2. Иностранные языки

**Условие.** На кафедре $13$ сотрудников, каждый знает минимум один из трёх языков. Английский знают $10$, немецкий $7$, французский $6$; английский и немецкий — $5$, английский и французский — $4$, немецкий и французский — $3$. Найти число знающих все три, ровно два и только английский.

**Решение.** Обозначим $x$ — число знающих все три. По включениям–исключениям:

$$13=10+7+6-5-4-3+x, \quad x=2.$$

Ровно два: $(5-2)+(4-2)+(3-2)=6$. Только английский: $10-5-4+2=3$.

**Ответ:** $2$; $6$; $3$.

In [ ]:
all_3=13-(10+7+6-5-4-3)
exactly_2=(5-all_3)+(4-all_3)+(3-all_3)
only_en=10-5-4+all_3
print('3.2:',all_3,exactly_2,only_en)
assert (all_3,exactly_2,only_en)==(2,6,3)

### Задача 3.3. Читательские вкусы

**Условие.** Журнал $A$ читают $60\%$, $B$ — $50\%$, $C$ — $50\%$; оба $A,B$ — $30\%$, $B,C$ — $20\%$, $A,C$ — $40\%$; все три — $10\%$. Найти доли не читающих ни одного, читающих ровно два и читающих не менее двух журналов.

**Решение.** Доля читающих хотя бы один: $60+50+50-30-20-40+10=80\%$; не читающих ни одного: $100-80=20\%$.

Ровно два: $(30-10)+(20-10)+(40-10)=60\%$. Не менее двух: $60+10=70\%$.

**Ответ:** $20\%$; $60\%$; $70\%$.

In [ ]:
union=60+50+50-30-20-40+10
none=100-union
exactly_2=(30-10)+(20-10)+(40-10)
at_least_2=exactly_2+10
print('3.3:',none,exactly_2,at_least_2)
assert (none,exactly_2,at_least_2)==(20,60,70)

## Беспорядки — задачи из практики № 3

Число перестановок $n$ различных предметов, при которых ни один не остаётся на своём месте:

$$D_n=n!\sum_{k=0}^n\frac{(-1)^k}{k!}=n!-C_n^1(n-1)!+C_n^2(n-2)!-\cdots+(-1)^n.$$

### Задача 4.1. Шляпы

**Условие.** Гардеробщица случайно раздаёт шесть шляп шести владельцам. Найти вероятности: а) все получают свою; б) никто не получает свою.

**Решение.** Всего $6!=720$ перестановок. Благоприятна одна для первого вопроса. Для второго $D_6=265$ беспорядков.

$$P_a=1/720, \qquad P_b=D_6/6!=265/720=53/144.$$

**Ответ:** $1/720$; $53/144$.

In [ ]:
p_hats_all=Fraction(1,factorial(6))
p_hats_none=Fraction(D(6),factorial(6))
print('4.1:',p_hats_all,p_hats_none);assert p_hats_none==Fraction(53,144)

### Задача 4.2. Проверка контрольных

**Условие.** $n$ студентов должны взаимно проверить контрольные, причём никто не проверяет свою. Сколько распределений возможно?

**Решение.** Это число беспорядков $n$ объектов. По включениям–исключениям вычитаем назначения, в которых хотя бы один студент получил свою работу:

$$D_n=\sum_{k=0}^n(-1)^k C_n^k(n-k)!=n!\sum_{k=0}^n\frac{(-1)^k}{k!}.$$

**Ответ:** $D_n$ по приведённой формуле; например, $D_5=44$.

In [ ]:
print('4.2: n=1..7:',{n:D(n) for n in range(1,8)})
assert D(5)==44

### Задача 4.3. Подарки

**Условие.** Пять членов семьи обменялись пятью подарками так, чтобы ни один не получил свой. Сколько распределений возможно?

**Решение.** Нужны беспорядки из пяти элементов:

$$D_5=5!-C_5^1\cdot4!+C_5^2\cdot3!-C_5^3\cdot2!+C_5^4\cdot1!-C_5^5\cdot0!$$
$D_5=120-120+60-20+5-1=44.$$

**Ответ:** $44$ способа.

In [ ]:
print('4.3:',D(5));assert D(5)==44

### Задача 4.4. Почтальон Печкин

**Условие.** Десять различных посылок доставлены наугад десяти адресатам. Найти вероятность того, что хотя бы один получит собственную посылку.

**Решение.** Всего $10!$ равновозможных распределений. Противоположное событие — беспорядок из десяти элементов. Поэтому:

$$P(\text{хотя бы одно совпадение})=1-\frac{D_{10}}{10!}=1-\frac{1\,334\,961}{3\,628\,800}\approx0{,}63212.$$

**Ответ:** $1-D_{10}/10!$.

In [ ]:
p_44=1-Fraction(D(10),factorial(10))
print('4.4: D10 =',D(10),'; P =',p_44,'≈',float(p_44))
assert D(10)==1334961

### Задача 4.5. Задача Монмора

**Условие.** Две колоды из $n$ различных карт сопоставляются карта за картой после случайного перемешивания одной из них. Найти вероятность, что не будет ни одного совпадения. В материалах практики число карт не уточнено, поэтому запишем общий ответ для $n$.

**Решение.** Считаем расположение первой колоды фиксированным. Всего $n!$ вариантов второй колоды, благоприятных $D_n$:

$P=\frac{D_n}{n!}=\sum_{k=0}^{n}\frac{(-1)^k}{k!}\approx e^{-1}\approx0{,}36787944.$

**Ответ:** точно $D_n/n!$; при обычной колоде из $52$ карт — около $0{,}36787944$.

In [ ]:
p_montmort=Fraction(D(52),factorial(52))
print('4.5:',float(p_montmort))
assert abs(float(p_montmort)-0.36787944117144233)<1e-12

# Лабораторная работа № 4
## Условная вероятность и формула полной вероятности

**Тема:** классическое решение и моделирование задачи с двумя урнами.

**Цель:** вычислить вероятность по формуле полной вероятности и проверить её методом Монте-Карло.

**Условие.** В первой урне 15 белых и 5 чёрных шаров, во второй — 6 белых и 1 чёрный. Из первой урны случайно вынимают 3 шара **без возвращения** и перекладывают во вторую. Затем из второй урны случайно вынимают 1 шар. Найти вероятность того, что последний шар белый.

### 1. Классическое решение

Пусть $A$ — событие «из второй урны вынут белый шар»; $H_k$ — гипотеза «из первой во вторую урну перенесено ровно $k$ белых шаров», $k=0,1,2,3$.

Всего способов выбрать 3 шара из первой урны: $C_{20}^{3}=1140$.

Вероятность гипотезы $H_k$ (гипергеометрическое распределение):

$$P(H_k)=\frac{C_{15}^{k}C_{5}^{3-k}}{C_{20}^{3}}.$$

После переноса во второй урне будет 10 шаров, среди которых $6+k$ белых. Поэтому

$$P(A\mid H_k)=\frac{6+k}{10}.$$

По **формуле полной вероятности**:

$$P(A)=\sum_{k=0}^{3}P(H_k)P(A\mid H_k).$$

In [1]:
from math import comb
from fractions import Fraction

ways_total = comb(20, 3)
probability = Fraction(0, 1)

print('Перенесено белых | Число способов | P(H_k) | P(A | H_k)')
for k in range(4):
    ways = comb(15, k) * comb(5, 3-k)
    p_h = Fraction(ways, ways_total)
    p_white_given_h = Fraction(6+k, 10)
    probability += p_h * p_white_given_h
    print(f'{k:^17} | {ways:^13} | {str(p_h):^8} | {p_white_given_h}')

print(f'\nТочная вероятность: {probability} = {float(probability):.6f}')
assert probability == Fraction(33, 40)


Перенесено белых | Число способов | P(H_k) | P(A | H_k)
        0         |      10       |  1/114   | 3/5
        1         |      150      |   5/38   | 7/10
        2         |      525      |  35/76   | 4/5
        3         |      455      |  91/228  | 9/10

Точная вероятность: 33/40 = 0.825000


Подставим все четыре гипотезы:

$$P(A)=\frac{10}{1140}\cdot\frac6{10}+\frac{150}{1140}\cdot\frac7{10}+\frac{525}{1140}\cdot\frac8{10}+\frac{455}{1140}\cdot\frac9{10}$$

$$=\frac{60+1050+4200+4095}{11400}=\frac{9405}{11400}=\frac{33}{40}=0{,}825.$$

**Ответ по классической формуле: $P(A)=0{,}825=82{,}5\%$.**

### 2. Проверка классического решения другим способом

После переноса во второй урне 10 шаров. Из них 7 первоначально находились во второй урне и 3 были перенесены из первой. Если вытянут шар из первоначальной части, шанс белого $6/7$; если перенесённый — $15/20$.

$$P(A)=\frac7{10}\cdot\frac67+\frac3{10}\cdot\frac{15}{20}=\frac6{10}+\frac9{40}=\frac{33}{40}.$$

In [2]:
check = Fraction(7, 10)*Fraction(6, 7) + Fraction(3, 10)*Fraction(15, 20)
print('Аналитическая проверка:', check, '=', float(check))
assert check == probability


Аналитическая проверка: 33/40 = 0.825


### 3. Моделирование методом Монте-Карло

Многократно воспроизводим условие задачи. В каждом испытании заново создаём обе урны, случайно перекладываем **три разных** шара из первой во вторую и выбираем один шар из второй. Отношение числа белых шаров к количеству испытаний даёт экспериментальную оценку вероятности.

Используем фиксированное начальное состояние генератора, чтобы результат воспроизводился. Сравним оценку после 100, 1000, 10 000 и 100 000 испытаний.

In [3]:
import random

rng = random.Random(20260928)
checkpoints = (100, 1_000, 10_000, 100_000)
white_outcomes = 0
estimates = []

for experiment in range(1, checkpoints[-1] + 1):
    first_urn = ['белый'] * 15 + ['чёрный'] * 5
    second_urn = ['белый'] * 6 + ['чёрный']

    transferred = rng.sample(first_urn, 3)  # без возвращения
    second_urn.extend(transferred)
    drawn = rng.choice(second_urn)
    if drawn == 'белый':
        white_outcomes += 1

    if experiment in checkpoints:
        estimate = white_outcomes / experiment
        estimates.append(estimate)
        print(f'{experiment:>6} испытаний: P ≈ {estimate:.5f}; '
              f'отклонение = {abs(estimate - float(probability)):.5f}')

assert abs(estimates[-1] - float(probability)) < 0.01


   100 испытаний: P ≈ 0.78000; отклонение = 0.04500
  1000 испытаний: P ≈ 0.83000; отклонение = 0.00500
 10000 испытаний: P ≈ 0.82110; отклонение = 0.00390
100000 испытаний: P ≈ 0.82404; отклонение = 0.00096


### 4. Сравнение результатов и вывод

**Теоретическая вероятность:** $33/40=0{,}825$.

**Модельная вероятность:** при 100 000 испытаний получено $0{,}82404$; при другом начальном состоянии генератора оценка немного изменится.

**Вывод:** классический расчёт и компьютерное моделирование согласуются. При большом числе независимых повторений относительная частота белого шара близка к вероятности $0{,}825$.